relay-sampling/src/lib.rs

```diff
@@ -427,57 +399,95 @@ impl SamplingRule {
     /// neither idle nor expired.
-    fn is_active(&self, now: DateTime<Utc>) -> bool {
-        self.time_range.contains(now)
-    }
-
-    /// Returns the sample rate of this specific sampling rule based on the existence of the
-    /// decaying function.
-    ///
-    /// In case of any problems that render the decaying function impossible to compute, we are
-    /// going to fallback to the base sample rate.
-    pub fn get_sample_rate(&self, now: DateTime<Utc>) -> f64 {
-        if let Some(params) = self.get_decaying_function_params(now) {
-            self.decaying_fn.call(params)
-        } else {
-            self.sample_rate
+    fn is_active(&self, now: DateTime<Utc>) -> Option<MatchingRule> {
+        match self.decaying_fn {
+            DecayingFunction::Linear {
+                decayed_sample_rate,
+            } => {
+                if let TimeRange {
+                    start: Some(start),
+                    end: Some(end),
+                } = self.time_range
+                {
+                    if self.sample_rate > decayed_sample_rate && start < end && now >= start {
+                        return Some(MatchingRule {
+                            id: self.id,
+                            evaluator: SampleRateEvaluator::Linear {
+                                start,
+                                end,
+                                sample_rate: self.sample_rate,
+                                decayed_sample_rate,
+                            },
+                        });
+                    }
+                }
+            }
+            DecayingFunction::Constant => {
+                if self.time_range.contains(now) {
+                    return Some(MatchingRule {
+                        id: self.id,
+                        evaluator: SampleRateEvaluator::Constant {
+                            sample_rate: self.sample_rate,
+                        },
+                    });
+                }
+            }
         }
-    }
 
-    /// Validates and gets the decaying function params.
-    fn get_decaying_function_params(&self, now: DateTime<Utc>) -> Option<DecayingFunctionParams> {
-        if let TimeRange {
-            start: Some(start),
-            end: Some(end),
-        } = self.time_range
-        {
-            if !self.validate_function_specific_params() || end <= start || now < start {
-                return None;
-            }
+        None
+    }
+}
 
-            let now_timestamp = now.timestamp() as f64;
-            let start_timestamp = start.timestamp() as f64;
-            let end_timestamp = end.timestamp() as f64;
-            let progress_ratio = ((now_timestamp - start_timestamp)
-                / (end_timestamp - start_timestamp))
-                .clamp(0.0, 1.0);
+/// A sampling rule that has been successfully matched and that contains all the required data
+/// to return the sample rate.
+#[derive(Debug, Clone, Copy)]
+pub struct MatchingRule {
```

Comment:
`ActiveRule` might actually be a better name now that I think of it, because it is created by `is_active`.

```suggestion
pub struct ActiveRule {
```